### PR TITLE
Dev

### PR DESCRIPTION
**Pull request by KennyTheBard**
Note:

- the database setup is with docker, not Heroku / MongoAtlas
- the API endpoints are with Hoppscotch, not Postman (you can find the collection in `hoppscotch.io.json`)
- the image uploading part is not implemented

### CODE REVIEW

File: hoppscotch.io.json [hunk 1]
```diff
@@ -0,0 +1,112 @@
+{
```
Comment (ionut-movila) [inline hoppscotch.io.json:1]: Nice, I did not knew about this alternative to Postman!

File: docker-compose.yml [hunk 2]
```diff
@@ -0,0 +1,19 @@
+version: '3.7'
```
Comment (ionut-movila) [inline docker-compose.yml:1]: We like Docker :)

File: package.json [hunk 3]
```diff
@@ -5,22 +5,25 @@
...
         "compile": "tsc",
         "compile:watch": "tsc -w",
-        "start:local": "nodemon -r dotenv/config build/src/index.js dotenv_config_path=.env.local",
+        "start:local": "nodemon -r dotenv/config build/src/index.js dotenv_config_path=.env",
         "start": "node build/src/index.js"
     },
     "author": "KennyTheBard",
     "license": "MIT",
     "dependencies": {
+        "axios": "^1.4.0",
```
Comment (ionut-movila) [inline package.json:14]: Maybe some more libs from here should have been included in the PR (that are not generated by npm init for instance). No need to change now, just a small detail.

File: src/config/index.ts [hunk 4]
```diff
@@ -0,0 +1,26 @@
+import { get } from "env-var";
+import * as dotenv from "dotenv";
+
+dotenv.config();
+
+export const appConfig: AppConfig = {
```
Comment (ionut-movila) [inline src/config/index.ts:6]: Clean and easy to understand 👍

File: src/index.ts [hunk 5]
```diff
@@ -0,0 +1,40 @@
+import mongoose from "mongoose";
+import { appConfig } from "./config";
+import cors from "cors";
+import express from "express";
+import { createPokemon, updatePokemon, deletePokemon, getOnePokemon, getAllPokemons, deleteAllPokemons, populateDatabase } from "./routes";
+import { PokemonService } from "./services";
+
+const bootstrap = async () => {
+    // init database connection
+    await mongoose.connect(
```
Comment (ionut-movila) [inline src/index.ts:10]: What is connect fails? Should we have a try/catch here and not continue further with the rest of the code?

File: src/models/Pokemon.ts [hunk 6]
```diff
@@ -0,0 +1,86 @@
+import mongoose, { Document, Schema } from "mongoose";
+import { ZodType, z } from "zod";
```
Comment (ionut-movila) [inline src/models/Pokemon.ts:2]: I like Zod lib, easy to use and light.

File: src/models/Pokemon.ts [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+export const IPokemonSchema = z.object({
+    id: z.number(),
+    name: z.string(),
+    height: z.number(),
+    weight: z.number(),
+    abilities: z.array(AbilitySchema),
+    heldItem: HeldItemSchema.or(z.null()),
+});
+
+const pokemonSchema = new Schema<IPokemon>({
```
Comment (ionut-movila) [inline src/models/Pokemon.ts:44]: I really like the detailed TS usage.

File: src/routes/index.ts [hunk 8]
```diff
@@ -0,0 +1,107 @@
+import { Request, Response } from "express";
+import Pokemon, { IPokemon, IPokemonSchema } from "../models/Pokemon";
+import { IdParamSchema } from "../utils";
+import { PokemonService } from "../services";
+
+export const createPokemon = async (req: Request, res: Response) => {
+    try {
+        const newPokemonData: IPokemon = IPokemonSchema.parse(req.body);
+        const newPokemon = await Pokemon.create(newPokemonData);
+        res.status(201).json(newPokemon);
```
Comment (ionut-movila) [inline src/routes/index.ts:10]: Maybe you can use a lib for status coded, not to have hardcoded numbers :D

File: src/services/PokemonApi.ts [hunk 9]
```diff
@@ -0,0 +1,126 @@
...
+
+export class PokemonService {
+    public refreshPokemons = async (): Promise<number> => {
+        const response = await axios.get(
+            "https://pokeapi.co/api/v2/pokemon?limit=151"
+        );
+        const pokemonEntries = response.data.results as PokemonListEntry[];
+
+        // TODO: could use some batching
+        const requests = await Promise.allSettled(
```
Comment (ionut-movila) [inline src/services/PokemonApi.ts:12]: How this compares to Promise.all? (is it better?)

File: src/utils/IdParamSchema.ts [hunk 10]
```diff
@@ -0,0 +1,9 @@
+import { z } from "zod";
+
+export const IdParamSchema = z
```
Comment (ionut-movila) [inline src/utils/IdParamSchema.ts:3]: Nice custom validation for the ID!

**Review by ionut-movila — APPROVED** [review]
Code looks clean and easy to understand, I like it! 💯

File: src/index.ts (hunk 5 above)
Comment (KennyTheBard) [reply src/index.ts:10]: if it fails with error, i'm OK with it crashing the app

No point in running without a DB connection :sweat_smile:

File: src/services/PokemonApi.ts (hunk 9 above)
Comment (KennyTheBard) [reply src/services/PokemonApi.ts:12]: It does the same thing for the most part, but doesn't throw on the first rejected promise
It's great if you can still use the resolved results, while handling (or ignoring) the failed ones afterwards

**Pull request merged**